### ISSUE TITLE
[CSharp] Add optimization of RepeatedField<T>.AddRange for ICollection<T>

### ISSUE THREAD

**Issue by vicocz** [protocolbuffers/protobuf#10769]
**What language does this apply to?**
If it's a proto syntax change, is it for proto2 or proto3?
No syntax change.
If it's about generated code change, what programming language?
C#

**Describe the problem you are trying to solve.**
`AddRange` method already has optimization when adding items from non-generic `ICollection` but the generic one is missing. If a caller adds items from e.g. `Dictionary<T>.Keys`, which is of `ICollection<T>`, items are added one by one causing incremental reallocation of its internal array. Especially if large number of items are being added.

**Describe the solution you'd like**
`ICollection<T>` should be handled in the same way is `ICollection` so as reallocation is performed once.

**Describe alternatives you've considered**
Yeah, enforce added values are of non-generic `ICollection` type, but it requires additional casting for a consumer which is kind of step back.

**Additional context**
Enhancement itself is expected to be quite simple, in fact copy pasted for `ICollection` part:
```
            if (values is ICollection<T> collection)
            {
                var extraCount = collection.Count;

                EnsureSize(count + extraCount);
                collection.CopyTo(array, count);
                count += extraCount;
                return;
            }
```
The only question is whether it generic enough as it does not cover e.g. `IReadOnlyCollection<T>` and other similar types.

**Comment by jskeet** [comment]
>  If a caller adds items from e.g. `Dictionary<T>.Keys`, which is of `ICollection<T>`, items are added one by one causing incremental reallocation of its internal array.

It shouldn't - because while `Keys` has a *compile-time* type of `ICollection<T>`, the actual type (`Dictionary<TKey, TValue>.KeyCollection`) *does* implement `ICollection`. Sample test code:

```csharp
using System;
using System.Collections;
using System.Collections.Generic;

var dict = new Dictionary<string, string> { { "a", "b" }};

Console.WriteLine(dict.Keys.GetType());   // System.Collections.Generic.Dictionary`2+KeyCollection[System.String,System.String]
Console.WriteLine(dict.Keys is ICollection); // True
```

There's a comment within the code to explain that:

```csharp
// We *could* check for ICollection<T> as well, but very very few collections implement
// ICollection<T> but not ICollection. (HashSet<T> does, for one...)
```

So while it's definitely feasible, I believe it's an optimization which would *relatively* rarely have an impact. I'm happy to look at it at some point, but I consider it relatively low priority. (I'm effectively at the level of a community contributor; this isn't my main role - so I'm afraid there's quite a long wait time for this sort of thing.)

**Comment by jskeet** [comment]
@vicocz: Any comment on this, given my observation about `Dictionary.Keys`?

**Comment by vicocz** [comment]
> Also note that if we _did_ do this, we'd need the same kind of per-element nullity check that we have for `ICollection`.

Yes that's clear, I just wanted to have a smallest possible code snippet.

**Comment by jskeet** [comment]
@vicocz: I'd like to close this due to my previous reasoning - I don't see this having enough benefit to make it worth the extra complexity. Is that okay with you? If you've got evidence that this would have a measurable benefit, we could look at it again, but otherwise I think it's best left as-is.

I'll close the issue now, but feel free to comment and we can reopen.
